plugins/CoreAdminHome/Tasks.php

```diff
@@ -258,6 +262,48 @@ public function updateSpammerBlacklist()
         Option::set(ReferrerSpamFilter::OPTION_STORAGE_NAME, serialize($list));
     }
 
+    public function purgeOrphanedArchives()
+    {
+        $segmentHashesByIdSite = $this->getSegmentHashesByIdSite();
+        $archiveTables = ArchiveTableCreator::getTablesArchivesInstalled('numeric');
+
+        $datesPurged = array();
+        foreach ($archiveTables as $table) {
+            $date = ArchiveTableCreator::getDateFromTableName($table);
+            list($year, $month) = explode('_', $date);
+
+            $dateObj = Date::factory("$year-$month-15");
+
+            $this->archivePurger->purgeDeletedSiteArchives($dateObj);
+            $this->archivePurger->purgeDeletedSegmentArchives($dateObj, $segmentHashesByIdSite);
+
+            $datesPurged[$date] = true;
+        }
+    }
+
+    /**
+     * Get a list of all segment hashes that currently exist, indexed by idSite.
+     * @return array
+     */
+    public function getSegmentHashesByIdSite()
+    {
+        //Get a list of hashes of all segments that exist now
+        $sql = "SELECT DISTINCT definition, enable_only_idsite FROM " . Common::prefixTable('segment')
+            . " WHERE deleted = 0";
+        $rows = Db::fetchAll($sql);
+        $segmentHashes = array();
+        foreach ($rows as $row) {
+            try {
+                $segment = new Segment($row['definition'], array());
```

Comment:
I'm quite certain we will need to pass the `idSites` here. Otherwise it won't find segments that only exist for a specific site. Eg `dimension1` is a segment for custom dimensions and it will be only valid segment if the idSite is passed.  

You can probably reproduce this by following these steps (haven't tested it)

* Creating a custom dimension
* Creating a segment that uses the created custom dimension
* running this command using eg `./console core:run-scheduled-tasks "Piwik\Plugins\CoreAdminHome\Tasks.purgeOrphanedArchives" --force`.

If `idSite=0`, then it should be fine to not pass any site parameter AFAIK.

Reply:
Once you pass `idSite` it should work like `array($row['idSite'])` but you may need to update the `CustomDimensions` submodule and fetch the latest updates in `master` as I fixed an issue around this just the other day.

Reply:
Have added a unit test for this case - now skipping segment validation anyway so this is no longer an issue